### ISSUE TITLE
Filtering not working

### ISSUE THREAD

**Issue by Blackdread** [jhipster/generator-jhipster#7195]
### The issue

Filtering on entities is not fully implemented based on [specifications](http://www.jhipster.tech/entities-filtering/)

For a OneToOne or ManyToOne relation, filters like Equals, In and IsSpecified will work but not GreaterThan, LessThan, GreaterOrEqualThan, LessOrEqualThan.

For a normal field (not a foreign key) (example id), filters like Equals, In and IsSpecified will work but not GreaterThan, LessThan, GreaterOrEqualThan, LessOrEqualThan.

Service class "**QueryService**" of jHipster does not filter all cases greaterThan, lessThan, etc.

Only dates and times works with greaterThan, lessThan, etc, it calls the method "buildRangeSpecification"

### Motivation for or Use Case

All filtering should work for entity fields. It is not normal to not be able to to filter on primary key ID via GreaterThan etc.
It should also allow by default to filter for ManyToOne and OneToOne with GreaterThan, LessThan, GreaterOrEqualThan, LessOrEqualThan.

### Reproduce the error

A simple app with OneToOne, ManyToMany, ManyToOne relations and fields then try to filter id or a referring entity with greaterThan or In or any other that is not supported.

### Related issues

None

### Suggest a Fix

Add more if/else to handle other filtering or use the method "buildRangeSpecification" for types that support range filtering (like int, long, etc)/

### The JDL file

entity Camera {
	name String required minlength(2) maxlength(120)
}

entity CameraParam {
	param String required maxlength(45)
}

relationship OneToOne {
	CameraParam{camera(name) required} to Camera{cameraParam required}
}

---

### Versions 
| JHipster           |
| ------------------ |
| `4.14.1` |

- [x ] Checking this box is mandatory (this is just to show you read everything)

### Code from QueryService

Helper function to return a specification for filtering on one-to-one or many-to-one reference
```
protected <OTHER, X> Specification<ENTITY> buildReferringEntitySpecification(Filter<X> filter,
        SingularAttribute<? super ENTITY, OTHER> reference,
        SingularAttribute<OTHER, X> valueField) {
        if (filter.getEquals() != null) {
            return equalsSpecification(reference, valueField, filter.getEquals());
        } else if (filter.getIn() != null) {
            return valueIn(reference, valueField, filter.getIn());
        } else if (filter.getSpecified() != null) {
            return byFieldSpecified(reference, filter.getSpecified());
        }
        return null;
    }
```

Helper function to return a specification for filtering on one-to-many or many-to-many reference
```
protected <OTHER, X> Specification<ENTITY> buildReferringEntitySpecification(Filter<X> filter,
        SetAttribute<ENTITY, OTHER> reference,
        SingularAttribute<OTHER, X> valueField) {
        if (filter.getEquals() != null) {
            return equalsSetSpecification(reference, valueField, filter.getEquals());
        } else if (filter.getSpecified() != null) {
            return byFieldSpecified(reference, filter.getSpecified());
        }
        return null;
    }
```

**This one is called by the ID field for example by default (so no range filter)**
Helper function to return a specification for filtering on a single field, where equality, and null/non-null conditions are supported.
```
protected <X> Specification<ENTITY> buildSpecification(Filter<X> filter, SingularAttribute<? super ENTITY, X>
        field) {
        if (filter.getEquals() != null) {
            return equalsSpecification(field, filter.getEquals());
        } else if (filter.getIn() != null) {
            return valueIn(field, filter.getIn());
        } else if (filter.getSpecified() != null) {
            return byFieldSpecified(field, filter.getSpecified());
        }
        return null;
    }
```

**By default for Instant field type it will call this method**
Helper function to return a specification for filtering on a single Comparable, where equality, less than, greater than and less-than-or-equal-to and greater-than-or-equal-to and null/non-null conditions are supported.
```
protected <X extends Comparable<? super X>> Specification<ENTITY> buildRangeSpecification(RangeFilter<X> filter,
        SingularAttribute<? super ENTITY, X> field) {
        if (filter.getEquals() != null) {
            return equalsSpecification(field, filter.getEquals());
        } else if (filter.getIn() != null) {
            return valueIn(field, filter.getIn());
        }

        Specifications<ENTITY> result = Specifications.where(null);
        if (filter.getSpecified() != null) {
            result = result.and(byFieldSpecified(field, filter.getSpecified()));
        }
        if (filter.getGreaterThan() != null) {
            result = result.and(greaterThan(field, filter.getGreaterThan()));
        }
        if (filter.getGreaterOrEqualThan() != null) {
            result = result.and(greaterThanOrEqualTo(field, filter.getGreaterOrEqualThan()));
        }
        if (filter.getLessThan() != null) {
            result = result.and(lessThan(field, filter.getLessThan()));
        }
        if (filter.getLessOrEqualThan() != null) {
            result = result.and(lessThanOrEqualTo(field, filter.getLessOrEqualThan()));
        }
        return result;
    }
```

**Comment by jdubois** [comment]
It's not working for some types of objects -> can you update the title to be more explicit? It's a feature request not a bug

**Comment by Blackdread** [comment]
Not sure this is a feature request as:

> Public interface
> For each entity, you can enable filtering in the entity generator, and after, you can call your /api/my-entity GET endpoint with the following parameters :
> 
> - For each xyz field
>   - xyz.equals=someValue
> To list all the entities, where xyz equals to ‘someValue’
>   - xyz.equals=someValue,otherValue
> To list all the entities, where xyz equals to ‘someValue’ or ‘otherValue’
>   - xyz.specified=true
> To list all the entities, where xyz is not null, specified.
>   - xyz.specified=false
> To list all the entities, where xyz is null, unspecified.
> - If xyz’s type is string:
>   - xyz.contains=something
> To list all the entities, where xyz contains ‘something’.
> - If xyz’s is either any of the number types, or the date types.
>   - xyz.greaterThan=someValue
> To list all the entities, where xyz is greater than ‘someValue’.
>   - xyz.lessThan=someValue
> To list all the entities, where xyz is less than ‘someValue’.
>   - xyz.greaterOrEqualThan=someValue
> To list all the entities, where xyz is greater than or equal to ‘someValue’.
>   - xyz.lessOrEqualThan=someValue
> - To list all the entities, where xyz is less than or equal to ‘someValue’.
> Of course, they can be combined freely.

From the quote, we see that any number (int, Long, Float, Double, BigDecimal etc) should support Range filtering and not only Equals, In or IsSpecified. This same behaviour should also be applied to foreign keys (referring entity) as FK are number (a Long for the case MySQL and JPA generation)

I have not checked String filtering if all works or not.

**Comment by Blackdread** [comment]
3 points:
- Calling **buildRangeSpecification** instead of **buildSpecification** for the ID (Long) field of the entity fixes the Range issue for that field

- Renaming methods **_buildSpecification, buildStringSpecification, buildRangeSpecification, buildReferringEntitySpecification, buildReferringEntitySpecification_** to **_buildSpecification_** works.
Runtime will pick appropriate methods based on filter/field/reference/valueField.
In **QueryService<ENTITY>**, it would be possible to add 5 identical methods that just call **_buildSpecification, buildStringSpecification, buildRangeSpecification, buildReferringEntitySpecification, buildReferringEntitySpecification_**; 
Generator would use by default the new methods and if user wants to specifically call a particular method for a field then he can modify the generated code to call any of the **_buildSpecification, buildStringSpecification, buildRangeSpecification, buildReferringEntitySpecification, buildReferringEntitySpecification_**.

- I have not tested yet to override **buildReferringEntitySpecification** to fix issue that Range filtering is not working for both of the methods but it should be feasible

**Comment by DanielFran** [comment]
@gzsombor I was expecting to PR sometime ago changes for same issue identified by @Blackdread....
Can you validate this approach?

Would love to see it in v5 asap.
Thanks

**Comment by gzsombor** [comment]
I always think primary key as an opaque token, which is 4 or 8 bytes long, and that's why I haven't added the range filtering for PKs. Definitely it's feasible, but I wondering what's the 'business case' - I always ended up with problems, when the primary key had a user specific meaning. Even when the users do need an increasing list of number for identifying their entities.
 Sure, it's not that hard to add support for it - the code is generic enough :)

**Comment by jdubois** [comment]
Filtering against the PKs is definitely a very weird idea - we shouldn't do this by default, let's not push people to do that kind of thing.

**Comment by DanielFran** [comment]
IMHO only in case of composite keys, that is the use case I have.

**Comment by jdubois** [comment]
Yes but composite keys are also a bad practice in my opinion -> we should allow people to do that, but not generate those by default

**Comment by DanielFran** [comment]
Of course, I agree with you. This just for existing database...

**Comment by Blackdread** [comment]
@DanielFran I would prefer not do the PR, I have not read that much the code in generator files "*.js"/"*.ejs". I checked the java part and generated code.

@gzsombor A primary key if auto-generated has never business logic and should never change. Only composite key can have but I would rather use an auto-generated one and add a unique constraint on my business logic as it simplifies a lot for joins etc.
Like @jdubois said we should not generate composite keys. Already many issues were created to get that feature but this is not really the best option (joins, etc).

The idea to give full filtering on primary key and foreign key (referring entity) is simply as it might be necessary to just want ranged IDs/FKs filtering, and it is confusing to not have range filtering on it when docs specify it should have (it is a number after all -> LongFilter).
Also swagger will generate the API documentation showing that ID and FKs can be filtered by range but in back-end it is not.
Also clients have access to the API and see of course the id keys and foreign keys, they might try to filter on it via range but unexpected behavior will happen for range filtering.



Just a note if that's of any interest for anyone (might be useful only for me but who knows), I have written a very simple java code to parse a SQL database and transform it into JDL. Had to do that as I had a lot of databases in production to change to jHipster apps and already all defined designs in MySQL Workbench tool. There is a lot of manual step even after the parser extracted the JDL ([here](https://github.com/Blackdread/sql-to-jdl))

**Comment by gzsombor** [comment]
The reverse engineering script would have some interest from others too, I believe. 😉 
I don't think, if JHipster can generate code which could work with composite keys, so I'm curious how your tool works.

You can override locally the buildReferringEntitySpecification methods of 'QueryService' - that methods are 'protected' - if you really need the functionality. 
 In my opinion,  I would rather change the swagger definition, to remove that confusing filters for the relations.

**Comment by Blackdread** [comment]
My tool does not consider composite keys as jhipster does not provide this feature, it is seen as normal foreign keys, and it will create corresponding ManyToOne relationships in the JDL file generated.

Could you provide the implementation of the buildReferringEntitySpecification for XToOne and ManyToMany with a RangeFilter<X> passed as parameter? Thank you for the help.

**Comment by Blackdread** [comment]
My pull request does not rename previous methods so it does not solve issue that we have to manually edit generated code to be able to add range filtering for ID field (primary Long).
What is the decision on it?
2 solutions:
1. change generator to use the **buildRangeSpecification** instead of **buildSpecification** for ID field
2. rename all to **buildSpecification** (string, range) but keep **buildReferringEntitySpecification**

3rd solution is to not do any above but document it on jhipster filtering page that ID field does not support range filtering by default but can by replacing method by  **buildRangeSpecification**

I am in favor of solution 2 but breaking change for previous apps generated and **loose possibility** to only call specific method as name are not different anymore (so solution 1 is better ^^ but I cannot do a pull request for that).
If solution 2 is not good then solution 1 would perfect and would follow what my pull request does with buildReferringEntitySpecification

**Comment by deepu105** [comment]
JHipster 5 can have breaking changes, so if its a cleaner solution then I
would vote for it but  @gzombor is the boss here


Thanks & Regards,
Deepu

**Comment by Blackdread** [comment]
Pull request [#49](https://github.com/jhipster/jhipster/pull/49) is merged, it fixes referring entities range filtering.
Only id does not support range filtering by default, same solutions possible as said in my previous post.

**Comment by deepu105** [comment]
@Blackdread so what is left to do for this? can you summarize

**Comment by Blackdread** [comment]
On id field cannot be filtered with ranges because it uses buildSpecification instead of range specificaion as id is a Long.
The rest is fixed as it will use "buildReferringEntitySpecification(RangeFilter<X> filter," if type of filter is compatible with RangeFilter (methods overloading)

**Comment by DanielFran** [comment]
I vote 2

**Comment by gzsombor** [comment]
I've opened a PR about a query service refactor, POC, here:  jhipster/jhipster#97. If you have time to review, and play around, give feedback about, it would be helpful, thanks.
